File: 443_StringCompression.js
/**
 * @param {character[]} chars
 * @return {number}
 */
var compress = function(chars) {
    //Runtime: 64 ms, faster than 76.90% of JavaScript online submissions for String Compression.
    //Memory Usage: 37.6 MB, less than 100.00% of JavaScript online submissions for String Compression.
    if(chars===[]){return 0};
    var newI=1,pre=chars[0],pre_count=1;
    for(i=1;i<chars.length;i++){
        if(pre!=chars[i]){
            pre=chars[i]
            if (pre_count>1){
                var tmp= pre_count.toString();
                for(var j=0;j<tmp.length;j++){
                    chars[newI++]=tmp[j]
                };
            };
            chars[newI]=chars[i]
            newI++
            pre_count=1

        }else{
            pre_count++
        }
    };
    if(pre_count>1){
        var tmp= pre_count.toString();
        for(var j=0;j<tmp.length;j++){
                chars[newI++]=tmp[j]
        };
    };

    while(chars.length>newI){chars.pop()};
    return newI;
};

var in_list = ["a","a","b","b","c","c","c"];
console.log(in_list);
console.log(compress(in_list));
console.log('-----------the end!----------------');

File: 443_StringCompression.py
class Solution(object):
    def compress(self, chars):
        """
        :type chars: List[str]
        :rtype: int
        """
        #Runtime: 32 ms, faster than 98.95% of Python online submissions for String Compression.
        #Memory Usage: 11.7 MB, less than 100.00% of Python online submissions for String Compression.
        # time: o(n) space:o(1)
        if not chars: return 0
        newI,preI,pre,pre_count=1,1,chars[0],1
        for i in range(1,len(chars)):
            print('i:'+str(i) +'  chars[i]:'+str(chars[i]))
            print('count:'+str(pre_count))
            print('newI:'+str(newI))

            if chars[i]!=pre:
                pre=chars[i]
                if pre_count>1:
                    tmp=str(pre_count)
                    chars[newI:newI+len(tmp)]=tmp
                    newI+=len(tmp)

                chars[newI]=chars[i]
                newI+=1
                pre_count=1
            else:
                pre_count+=1
        if pre_count>1:
            tmp=str(pre_count)
            chars[newI:newI+len(tmp)]=tmp
            newI+=len(tmp)

        while len(chars)>newI:chars.pop()
        print chars
        return newI

   def compress(self, chars):
        """
        :type chars: List[str]
        :rtype: int
        """
        # Runtime: 63 ms, faster than 43.03% of Python online submissions for String Compression.
        # Memory Usage: 13.9 MB, less than 17.26% of Python online submissions for String Compression.
        # time: O(n) space: o(1)
        # Add a end mark char in the end of the string to easy the handling of loop
        chars += [' ']
        # cur: cur char in the done list
        # cnt: current count of the cur char
        # index: the cur position to be filled by new context
        cur, cnt, index = chars[0], 1, 0
        for i in range(1, len(chars)):
            if chars[i] == cur:
                cnt += 1
            else:
                if cnt == 1:
                    chars[index] = cur
                    cur = chars[i]
                    index += 1
                else:
                    chars[index] = cur
                    index += 1
                    tmp = str(cnt)
                    chars[index:index+len(tmp)] = tmp
                    index += len(tmp)
                    cnt = 1
                    cur = chars[i]
        while len(chars) > index:
            chars.pop()
        return index


if __name__ == '__main__':
    object = Solution()
    num =  ["b","b","b","b","b","b","b","b","b","b","b","a","a","c"]
    #num=["a","b","b","b","b","b","b","b","b","b","b","b","b"]
    #num="abab"
    #p="ab"

    print(num)
    print('---Start---')
    r = object.compress(num)
    print(r)
    print('---End---')
